### PR TITLE
Splunk debug version

### PR DESCRIPTION
**Pull request by guykeller**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/27106

## Description
Splunk is missing notable incidents intermittently.
This problem is now fixed.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Demisto
- [ ] 5.0.0
- [ ] 5.5.0
- [ ] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py [hunk 1]
```diff
@@ -197,8 +204,11 @@ def notable_to_incident(event):
         incident["details"] = event["rule_description"]
     if demisto.get(event, "_time"):
         incident["occurred"] = event["_time"]
+        demisto.debug()
```
Comment (amshamah419) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py:207]: You passing something to be debugged?

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py [hunk 2]
```diff
@@ -497,15 +565,50 @@ def fetch_incidents(service):
...
+                                                                     len(incidents), incidents)
+    opt_in_log(debug_message)
+    latest_incident_fetched_time = None if len(incidents) == 0 else get_latest_incident_time(incidents)
+
+    fetches_with_same_start_time_count = demisto.getLastRun().get('fetch_start_update_count', 0) + 1
 
     demisto.incidents(incidents)
-    if len(incidents) < FETCH_LIMIT:
-        demisto.setLastRun({'time': now, 'offset': 0})
+    opt_in_log('SplunkPy - Found incidents at the end of this run: {}'.format(last_run_fetched_ids))
```
Comment (yuvalbenshalom) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py:596]: this are not the incident found in this run, it include incidents from other runs that were not cleaned. is this intended?

**Review by yuvalbenshalom — CHANGES_REQUESTED** [review]
lets review this togetgether and walk me thru the algorithm so I Weill be sure I understand.
written some easy fixes but mostly questions about how we want to do things, so lets talk about them.

we need to add unit tests for this fetch flow. it is complex and bound to be broken by next editors of the integration. need to think how to create good cases to cover it.

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.yml [hunk 3]
```diff
@@ -607,11 +655,7 @@ script:
     execution: false
     name: get-mapping-fields
   dockerimage: demisto/splunksdk:1.0.0.11989
-  feed: false
   isfetch: true
-  longRunning: false
-  longRunningPort: false
-  ismappable: true
```
Comment (yaakovi) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.yml:614]: why was this removed?

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py [hunk 4]
```diff
@@ -458,13 +469,63 @@ def splunk_results_command(service):
...
+
+
+def get_next_start_time(last_run, fetches_with_same_start_time_count, were_new_incidents_found=True):
+    last_run_datetime = occurred_to_datetime(last_run)
+    if were_new_incidents_found:
+        # Decreasing one minute to avoid missing incidents that were indexed late
+        last_run_datetime = last_run_datetime - timedelta(minutes=1)
+
+    # keep last time max 20 mins before current time, to avoid timeout
+    if fetches_with_same_start_time_count >= 20 and not were_new_incidents_found:  # TODO : make parameter
```
Comment (yuvalbenshalom) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py:496]: add pram?

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py [hunk 5]
```diff
@@ -458,13 +469,63 @@ def splunk_results_command(service):
...
+    next_run_without_miliseconds_and_tz = last_run_datetime.strftime(SPLUNK_TIME_FORMAT)
+    next_run = next_run_without_miliseconds_and_tz
+    return next_run
+
+
+def create_incident_custom_id(incident):
+    incident_raw_data = json.loads(incident['rawJSON'])['_raw']
+    incident_occurred = incident['occurred']
+    incident_custom_id = incident_occurred + incident_raw_data
+    incident_custom_id_hash = hashlib.md5(incident_custom_id).hexdigest()
```
Comment (yuvalbenshalom) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py:508]:
```suggestion
    raw_hash=hashlib.md5(incident_raw_data).hexdigest()
    incident_custom_id_hash = '{date}_{hash}'.format(date=incident_occurred, hash=raw_hash) 
```
I think it will be better if id will include redouble time. it will be easier to debug.

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py [hunk 6]
```diff
@@ -458,13 +469,63 @@ def splunk_results_command(service):
...
+    demisto.debug('length of incident new custom ID is: {}'.format(len(incident_custom_id)))
```
Comment (yuvalbenshalom) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy.py:509]:
```suggestion
```
will be constant now

File: Packs/SplunkPy/Integrations/SplunkPy/SplunkPy_test.py [hunk 7]
```diff
@@ -468,3 +455,434 @@ def test_fetch_incidents(mocker):
...
+    Then:
+    - The IDs of the two incidents are unique.
+    """
+    from SplunkPy import create_incident_custom_id
+    first_incident_custom_id = create_incident_custom_id(first_incident)
+    second_incident_custom_id = create_incident_custom_id(second_incident)
+    assert first_incident_custom_id != second_incident_custom_id
+
+
+def test_create_incident_custom_id_length():
```
Comment (yuvalbenshalom) [inline Packs/SplunkPy/Integrations/SplunkPy/SplunkPy_test.py:536]: I think it is not relevant because its an md5 so you can just test it is 32 characters, or if you take mt suggestion, 32 + 1 + len(occurred) that should also be const.

**Review by yuvalbenshalom — COMMENTED** [review]
looks good see my comments.

**Comment by guykeller** [comment]
Closing this, opened a new PR
